lib.nix

```diff
@@ -0,0 +1,131 @@
+let
+  # Export a Nix value to be consumed by Nickel
+  exportForNickel = value:
+      let type = builtins.typeOf value; in
+      if (type == "set") then (
+        if (value.type or "" == "derivation") then
+          { type = "nixDerivation"; drvPath = value.drvPath; outputName = value.outputName; }
+        else
+          builtins.mapAttrs (_: exportForNickel) value
+        )
+      else if (type == "list") then builtins.map exportForNickel value
+      else if (type == "list") then
+        throw "Can’t export a function"
+      else value;
+
+  # Import a Nickel value produced by the Nixel DSL
+  importFromNickel = mkShell: value:
+    let
+      type = builtins.typeOf value;
+      isNickelDerivation = type: type == "nickelPackage" || type == "nickelShell";
+      importFromNickel_ = importFromNickel mkShell;
+    in
+    if (type == "set") then (
+      let valueType = value.type or ""; in
+      if valueType == "nickelPackage" then
+        derivation (builtins.mapAttrs (_: importFromNickel_) value)
+      else if valueType == "nickelShell" then
+        mkShell (builtins.mapAttrs (_: importFromNickel_) value)
+      else if valueType == "nixDerivation" then
+        (import value.drvPath).${value.outputName or "out"}
+      else if valueType == "nixString" then
+        builtins.concatStringsSep "" (builtins.map importFromNickel_ value.fragments)
+      else
+        builtins.mapAttrs (_: importFromNickel_) value
+      )
+    else if (type == "list") then
+        builtins.map importFromNickel_ value
+    else value;
+
+  # Generate a Nickel program that evaluates the nickel-nix output, passing
+  # the given exported packages, and write it to outFile.
+  computeNickelFile = system: {nickelFile, exportedPkgs}:
+    let
+      exportedJSON = builtins.toFile
+          "inputs.json"
+          (builtins.unsafeDiscardStringContext (builtins.toJSON (exportForNickel exportedPkgs)));
```

Comment:
This `unsafeDiscardStringContext` is mildly nasty. It would be cleaner to bring `nixpkgs` into scope and use `pkgs.writeText` (which allows bringing in a context) rather than `builtins.toFile`. But I guess depending on nixpkgs is annoying, so feel free to ignore me :wink:

Reply:
To be honest I just copy pasted to other code that exports thing. We already depend on a bunch of things, so it's fine to have one one more dependency. Could we just depend on `writeText` directly?

Reply:
I just tried and it indeed starts to bring in a lot of things, it seems. I guess the `stdenv` and all?

Reply:
Yes, you can’t really get it without pulling the whole world. So `toFile` will be enough for the time being :))